### PR TITLE
Get rid of unnecessary memmove() for MacOS.

### PR DESCRIPTION
**Pull request by dreamer-dead**
Hi.

Here is copy of PR in my own fork of Lightpack, https://github.com/dreamer-dead/MyLightpack/pull/2
I added one commit from my master branch except that to prettify MacOS grabber code.

Before applying this patch default profiler for Mac(Instruments) tells me that intrinsic memove() call eats about 80% of all CPU consumed by the Prismatic app even in Release configuration (https://yadi.sk/i/1AS6f0eDuZr8n). That's why I tried to remove it.
Please, take a look at this PR.

Note, that Mac build seems broken on my machine with Qt5.5 and Clang compiler due to missing flags `QMAKE_CXXFLAGS += -std=c++11 -stdlib=libc++`. MSVS enables C++1x mode by default.

### CODE REVIEW

**Comment by dreamer-dead** [comment]
Other grabbers still use `imgData` and pass it to `memcpy()` and `free()` calls, so it shouldn't be `const *`.
Will fix other platform builds.

**Comment by psieg** [comment]
ok to test

**Comment by dreamer-dead** [comment]
Builds is OK here.
@psieg, could you take a look?

**Comment by psieg** [comment]
Will do. I don't have time this weekend though, so next week

File: Software/grab/GrabberBase.cpp [hunk 1]
```diff
@@ -174,7 +174,11 @@ void GrabberBase::grab()
             const int bytesPerPixel = 4;
             QRgb avgColor;
             if (_context->grabWidgets->at(i)->isAreaEnabled()) {
-                Calculations::calculateAvgColor(&avgColor, grabbedScreen->imgData, grabbedScreen->imgFormat, grabbedScreen->screenInfo.rect.width() * bytesPerPixel, preparedRect );
+                Q_ASSERT(grabbedScreen->imgData);
+                Calculations::calculateAvgColor(
+                    &avgColor, grabbedScreen->imgData, grabbedScreen->imgFormat,
+                    grabbedScreen->screenInfo.rect.width() * bytesPerPixel,
+                    preparedRect );
```
Comment (psieg) [inline Software/grab/GrabberBase.cpp]: I know formatting is far from consistent, but if you change it, please remove the space before the closing bracket. I prefer having one line per argument (if multi line) for readability.

File: Software/grab/MacOSGrabber.cpp [hunk 2]
```diff
@@ -48,104 +122,49 @@ MacOSGrabber::~MacOSGrabber()
...
-    CGDataProviderRef provider = CGImageGetDataProvider(imageRef);
-    CFDataRef dataref = CGDataProviderCopyData(provider);
-    memcpy(screen->imgData, CFDataGetBytePtr(dataref), width * height * 4);
-    CFRelease(dataref);
 }
 
 bool MacOSGrabber::reallocate(const QList<ScreenInfo> &screens)
 {
-    const size_t kBytesPerPixel = 4;
+    const qreal pixelRatio = ((QGuiApplication*)QCoreApplication::instance())->devicePixelRatio();
```
Comment (psieg) [inline Software/grab/MacOSGrabber.cpp]: Does this also fix the retina grab offset issues?

File: Software/grab/include/GrabberBase.hpp [hunk 3]
```diff
@@ -40,26 +44,33 @@ enum GrabResult {
...
+
+#ifdef MAC_OS_CG_GRAB_SUPPORT
+    CGImageRef displayImageRef = nullptr;
+    CFDataRef imageDataRef = nullptr;
+
+    // This is only a view to the |imageDataRef| pixels raw data.
+    const unsigned char * imgData = nullptr;
+#else
+    unsigned char * imgData = nullptr;
+#endif
```
Comment (psieg) [inline Software/grab/include/GrabberBase.hpp]: The preferred (or rather designed) method of using this is the `associatedData` pointer to be used by the grabber. That is, create a custom struct/object in your grabber and point to it from associatedData. See `DDuplScreenData` for reference

Comment (dreamer-dead) [inline Software/grab/include/GrabberBase.hpp]: I agree, this version looks like a hack, with all these ifdefs.

File: Software/grab/GrabberBase.cpp (hunk 1 above)
Comment (dreamer-dead) [inline Software/grab/GrabberBase.cpp]: Will do.

**Comment by danielcompton** [comment]
@dreamer-dead have you had a chance to look at this further? It looks like it would make a big improvement to Prismatik on macOS.

**Comment by psieg** [comment]
@danielcompton hi, would you mind testing [this version](https://psieg.de/lightpack/osx_builds/Prismatik_5.11.2.11_129.dmg)? It's my adaptation of this PR, but I don't have a mac to test it.

**Comment by danielcompton** [comment]
That version crashes immediately on launch with the following error:

```
Termination Reason:    DYLD, [0x1] Library missing

Application Specific Information:
dyld: launch, loading dependent libraries

Dyld Error Message:
  Library not loaded: /usr/local/Cellar/qt@5.7/5.7.1/lib/QtGui.framework/Versions/5/QtGui
  Referenced from: /Users/danielcompton/Downloads/Prismatik.app/Contents/Frameworks/QtWidgets.framework/Versions/5/QtWidgets
  Reason: image not found
```

These ones were loaded though:

```
Binary Images:
       0x10d3a8000 -        0x10d508ff7 +de.psieg.Prismatik (0) <C82A3C6B-5812-3462-9266-6BC7D2601A63> /Users/USER/Downloads/Prismatik.app/Contents/MacOS/Prismatik
       0x10d572000 -        0x10da70ff7 +org.qt-project.QtWidgets (5.7 - 5.7.1) <07745F14-81CC-3129-BD1B-8BCD45CA26AC> /Users/USER/Downloads/Prismatik.app/Contents/Frameworks/QtWidgets.framework/Versions/5/QtWidgets
       0x10dbd3000 -        0x10e025fff +org.qt-project.QtGui (5.7 - 5.7.1) <BCD35856-390B-361F-9806-442DB90E8280> /Users/USER/Downloads/Prismatik.app/Contents/Frameworks/QtGui.framework/Versions/5/QtGui
       0x10e116000 -        0x10e635ff7 +org.qt-project.QtCore (5.7 - 5.7.1) <5E9BC277-7BA8-396A-B832-3008C55F8A89> /Users/USER/Downloads/Prismatik.app/Contents/Frameworks/QtCore.framework/Versions/5/QtCore
       0x10e6e4000 -        0x10e7ecfff +org.qt-project.QtNetwork (5.7 - 5.7.1) <8401DF4D-C78D-374A-8FB9-7E5776F71932> /Users/USER/Downloads/Prismatik.app/Contents/Frameworks/QtNetwork.framework/Versions/5/QtNetwork
       0x10e82c000 -        0x10e83aff7 +org.qt-project.QtSerialPort (5.7 - 5.7.1) <39E65C37-EF4A-371A-8F34-590796086241> /Users/USER/Downloads/Prismatik.app/Contents/Frameworks/QtSerialPort.framework/Versions/5/QtSerialPort
       0x10f5e2000 -        0x10f61fdc7  dyld (433.5) <8239D0D7-66F6-3C44-A77F-586F74525DA3> /usr/lib/dyld
```

**Comment by danielcompton** [comment]
Checking out https://github.com/psieg/Lightpack/pull/101 and building myself with `make` showed a marked improvement. Before CPU hovered around 25-35%, now it's sitting at 7%.

**Comment by psieg** [comment]
Merged via #101, thanks @dreamer-dead

**Pull request merged**